src/Locator/ChangedFiles.php

```diff
@@ -46,6 +46,66 @@ public function locateFromGitRepository(): FilesCollection
         return $this->parseFilesFromDiff($diff);
     }
 
+    public function getLocalRef(): ?string
+    {
+        return $this->repository->run('symbolic-ref', ['HEAD']);
+    }
+    public function getLocalSHA1(): ?string
+    {
+        return $this->repository->run('rev-parse', ['HEAD']);
+    }
+    private function getRemoteRef(): ?string
+    {
+        return $this->repository->tryToRunWithFallback(
+            function (): ?string {
+                return $this->repository->run('rev-parse', ['--abbrev-ref', '--symbolic-full-name', '@{u}']);
+            },
+            'null'
+        );
+    }
+    
+    public function getRemoteSHA1(): ?string
+    {
+        if ($this->getRemoteRef() === 'null') {
+            return null;
+        }
+        /**
+         * @psalm-suppress PossiblyUndefinedArrayOffset
+         * @psalm-suppress PossiblyNullArgument
+         */
+        list($remoteName, $branchName) = explode('/', $this->getRemoteRef(), 2);
+        $output = $this->repository->run('ls-remote', [$remoteName, $branchName]);
+        return $output ? strtok($output, "\t") : null;
+    }
+
+    /**
+     * @return FilesCollection
+     */
+    public function locateFromGitDiff(string $fromSha, ?string $toSha): FilesCollection
```

Comment:
What worries me a bit, is that there seem to be a lot of different states possible which require slightly different approaches / solutions.
How sure are you that all possible situations are covered?

Maybe the new PushedFiled locator should also contain details on how it works and what it covers exactly.